source/particles/particle_handler.cc

```diff
@@ -358,18 +418,34 @@ namespace Particles
     const Particle<dim, spacedim> &                                    particle,
     const typename Triangulation<dim, spacedim>::active_cell_iterator &cell)
   {
-    typename std::multimap<internal::LevelInd,
-                           Particle<dim, spacedim>>::iterator it =
-      particles.insert(
-        std::make_pair(internal::LevelInd(cell->level(), cell->index()),
-                       particle));
+    Assert(triangulation != nullptr, ExcInternalError());
 
-    particle_iterator particle_it(particles, it);
-    particle_it->set_property_pool(*property_pool);
+    if (particles.size() == 0)
+      particles.resize(triangulation->n_active_cells());
+
+    typename Triangulation<dim, spacedim>::active_cell_iterator cell_to_insert;
+
+    if (cell->is_locally_owned())
+      cell_to_insert = cell;
+    else
+      {
+        cell_to_insert = triangulation->begin_active();
+        while (cell_to_insert->is_locally_owned() == false)
+          ++cell_to_insert;
+
+        Assert(
+          cell_to_insert->is_locally_owned(),
+          ExcMessage(
+            "ParticleHandler<dim, spacedim>::insert_particle can only insert particles into locally owned cells, but none was found."));
+      }
+
+    const unsigned int active_cell_index = cell_to_insert->active_cell_index();
+    particles[active_cell_index].push_back(particle);
```

Comment:
Maybe we could think about sorting these vectors in the future? About how many particles in a cell are we talking about in general?

Reply:
I am not sure how you would sort them here. What would be the sorting criteria? The only useful criteria I could see is the order of property pool handles, so that all particle data is stored continuously in the order of particle iteration. However, that is a sort on all local particles, I would not want to do that every time a single particle is inserted/removed/moves to a different cell. Currently this happens during every mesh refinement/repartition.

Reply:
Maybe once globally?

Reply:
I agree, sorting things upon mesh refinement sounds reasonable unless we have indication about that with cache analysis data (because it is mostly memory fragmentation with poor prefetching I could imagine as a potential problem here).